srml/evm/src/lib.rs

```diff
@@ -0,0 +1,247 @@
+	/// Currency type for deposit and withdraw.
+	type Currency: Currency<Self::AccountId>;
+	/// The overarching event type.
+	type Event: From<Event> + Into<<Self as system::Trait>::Event>;
+	/// Precompiles associated with this EVM engine.
+	type Precompiles: Precompiles;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Example {
+		Accounts get(fn accounts) config(): map H160 => Account;
+		AccountStorages: double_map H160, blake2_256(H256) => H256;
+	}
+}
+
+decl_event!(
+	/// EVM events
+	pub enum Event {
+		/// Ethereum events from contracts.
+		Log(Log),
+	}
+);
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event() = default;
+
+		#[weight = SimpleDispatchInfo::FixedNormal(10_000)]
+		fn deposit_balance(origin, value: BalanceOf<T>) -> Result {
+			let sender = ensure_signed(origin)?;
+
+			T::Currency::withdraw(
+				&sender,
+				value,
+				WithdrawReason::Reserve,
+				ExistenceRequirement::KeepAlive,
+			)?;
+
+			let bvalue = U256::from(UniqueSaturatedInto::<u64>::unique_saturated_into(value));
+			let address = T::ConvertAccountId::convert_account_id(sender);
+			Accounts::mutate(&address, |account| {
+				account.balance += bvalue;
+			});
+
+			Ok(())
+		}
+
+		#[weight = SimpleDispatchInfo::FixedNormal(10_000)]
+		fn withdraw_balance(origin, value: BalanceOf<T>) -> Result {
+			let sender = ensure_signed(origin)?;
+			let address = T::ConvertAccountId::convert_account_id(sender.clone());
+			let bvalue = U256::from(UniqueSaturatedInto::<u64>::unique_saturated_into(value));
+
+			if Accounts::get(&address).balance < bvalue {
+				return Err("Not enough balance to withdraw")
+			}
+
+			Accounts::mutate(&address, |account| {
+				account.balance -= bvalue;
+			});
+
+			T::Currency::deposit_creating(&sender, value);
+
+			Ok(())
+		}
+
+		#[weight = SimpleDispatchInfo::FixedNormal(10_000)]
+		fn call(origin, target: H160, input: Vec<u8>, value: U256, gas_limit: u32) -> Result {
+			let sender = ensure_signed(origin)?;
+			let source = T::ConvertAccountId::convert_account_id(sender.clone());
+			let gas_price = T::FeeCalculator::gas_price();
+
+			let vicinity = Vicinity {
+				gas_price,
+				origin: source,
+			};
+
+			let mut backend = Backend::<T>::new(&vicinity);
+			let mut executor = StackExecutor::new_with_precompile(
+				&backend,
+				gas_limit as usize,
+				&backend::GASOMETER_CONFIG,
+				T::Precompiles::execute,
+			);
+
+			let total_fee = gas_price * U256::from(gas_limit);
+			if Accounts::get(&source).balance < value + total_fee {
+				return Err("Not enough balance to pay transaction fee")
+			}
+			executor.withdraw(source, total_fee).map_err(|_| "Withdraw fee failed")?;
+
+			let reason = executor.transact_call(
+				source,
+				target,
+				value,
+				input,
+				gas_limit as usize,
+			);
+
+			let ret = match reason {
+				ExitReason::Succeed(_) => Ok(()),
+				_ => Err("Execute message call failed")
```

Comment:
is it possible to get more detailed reasons?

Reply:
I added different error messages for all four return reason. Normal exit `Succeed`, Common out-of-gas errors `Error`, Explicit revert `Revert` and VM fatal error (such as operation not supported) `Fatal`.